File: swift/inventory.swift
import Foundation

struct InventoryItem {
    var id: String
    var name: String
    var price: Float
}

enum StoreEventType {
    case PURCHASED
    case REFUNDED
}

struct StoreEvent {
    var eventType: StoreEventType
    var timestamp: Date? = Date()
    var payload: Any
}

class Store {
    var name: String
    var openingTime: Int?
    var closingTime: Int?
    var stock: [String: Int] = [:]
    var events: [StoreEvent] = []

    init(name: String) {
        self.name = name
    }
    
    func updateItemQuantity(item: InventoryItem, amount: Int) {
        self.stock[item.id] = amount
    }

    func purchaseItem(item: InventoryItem) {
        guard self.stock[item.id] ?? 0 > 0 else { return }
        self.events.append(StoreEvent(eventType: StoreEventType.PURCHASED, payload: item.id))
        self.updateItemQuantity(item: item, amount: self.stock[item.id]! - 1)
    }

    func refundItem(item: InventoryItem) {
        self.events.append(StoreEvent(eventType: StoreEventType.REFUNDED, payload: item.id))
        self.updateItemQuantity(item: item, amount: (self.stock[item.id] ?? 0) + 1)
    }
}

let item1 = InventoryItem(id: "1", name: "Item 1 - A", price: 500)
let item2 = InventoryItem(id: "2", name: "Item 2 - B", price: 750)
let item3 = InventoryItem(id: "3", name: "Item 3 - C", price: 1000)

var store = Store(name: "Bob's Shop")

store.updateItemQuantity(item: item1, amount: 10)
store.updateItemQuantity(item: item2, amount: 7)
store.updateItemQuantity(item: item3, amount: 5)

store.purchaseItem(item: item1)

print(store.stock)

File: php/inventory.php
<?php

namespace Inventory;

trait Exposable {
    public function __get($key)
    {
        return $this->$key;
    }

    public function __set($key, $value)
    {
        $this->$key = $value;
    }
}


class InventoryItem
{
    use Exposable;
    
    private $id;
    private $name;
    private $price;

    public function __construct($id, $name, $price)
    {
        $this->id = $id;
        $this->name = $name;
        $this->price = $price;
    }
}

enum StoreEventType
{
    case PURCHASED;
    case REFUNDED;

    public function text()
    {
        return match ($this)
            {
                self::PURCHASED => 'PURCHASED',
                self::REFUNDED => 'REFUNDED'
            };
    }
}

class StoreEvent
{
    use Exposable;
    
    private $eventType;
    private $timestamp;
    private $payload;

    public function __construct($eventType, $payload)
    {
        $this->eventType = $eventType;
        $date = new \DateTimeImmutable();
        $this->timestamp = $date->getTimestamp();
        $this->payload = $payload;
    }
}

class Store
{
    use Exposable;
    
    private $name;
    private $openingTime;
    private $closingTime;
    private $stock;
    private $events;

    public function __construct($name)
    {
        $this->name = $name;
        $this->openingTime = 900;
        $this->closingTime = 1700;
        $this->stock = [];
        $this->events = [];
    }

    public function updateItemQuantity($item, $amount)
    {
        $this->stock[$item->id] = $amount;
    }

    public function purchaseItem($item)
    {
        if ($this->stock[$item->id] <= 0) return;
        $this->events[] = new StoreEvent(StoreEventType::PURCHASED, $item->id);
        $this->updateItemQuantity($item, $this->stock[$item->id] - 1);
    }

    public function refundItem($item)
    {
        $this->events[] = new StoreEvent(StoreEventType::REFUNDED, $item->id);
        $this->updateItemQuantity($item, $this->stock[$item->id] + 1);
    }
}

$item1 = new InventoryItem('1', 'Item 1 - A', 500);
$item2 = new InventoryItem('2', 'Item 2 - B', 750);
$item3 = new InventoryItem('3', 'Item 3 - C', 1000);

$items = [
    $item1->id => $item1,
    $item2->id => $item2,
    $item3->id => $item3
];
    

$store = new Store("Bob's Shop");

$store->updateItemQuantity($item1, 10);
$store->updateItemQuantity($item2, 7);
$store->updateItemQuantity($item3, 5);

$store->purchaseItem($item1);
$store->purchaseItem($item2);
$store->refundItem($item1);

foreach ($store->events as $event)
{
    echo "{$event->timestamp} {$event->eventType->text()} '{$items[$event->payload]->name}' \n";
}

?>
